Mathlib/LinearAlgebra/QuadraticForm/Basis.lean

```diff
@@ -91,4 +91,79 @@ noncomputable def toBilinHom (bm : Basis ι R M) : QuadraticMap R M N →ₗ[S]
   map_add' := add_toBilin bm
   map_smul' := smul_toBilin S bm
 
+open Finsupp in
+theorem map_finsupp_sum (Q : QuadraticMap R M N) (f : ι →₀ R) (g : ι → R → M) :
+    Q (f.sum g) = (f.sum fun i r => Q (g i r)) +
+    ∑ p ∈ Finset.filter (fun p ↦ p.1 < p.2) f.support.offDiag,
+      (polar Q) (g p.1 (f p.1)) (g p.2 (f p.2)) := by
```

Comment:
I'd be inclined to introduce a `Finsupp.sym2OffDiag` here, so that this theorem holds without `LinearOrder ι`

Reply:
Would that be something of the form `def _root_.Finsupp.sym2OffDiag (f : ι →₀ R) : Finset (ι × ι) := sorry`?

Reply:
Perhaps `def Finsupp.sym2OffDiag (f : ι →₀ R) : Set (Sym2 ι) := { s : Sym2 ι | ¬ s.IsDiag }`?

Reply:
Perhaps
````
def Finsupp.sym2OffDiag (f : ι →₀ R) : Set (Sym2 ι) := 
  Sym2.mk '' (f.support ×ˢ f.support) ∩ { s : Sym2 ι | ¬ s.IsDiag }
```
?

Reply:
Then the statement would be:
```
noncomputable instance : DecidableEq ι := by exact Classical.typeDecidableEq ι

open Finsupp in
theorem map_finsupp_sum (Q : QuadraticMap R M N) (f : ι →₀ R) (g : ι → R → M) :
    Q (f.sum g) = (f.sum fun i r => Q (g i r)) +
    ∑ p ∈ f.sym2OffDiag,
      Sym2.lift ⟨fun i j => (polar Q) (g i (f i)) (g j (f j)), fun _ _ => polar_comm _ _ _⟩ p := sorry
```

Reply:
(I'm stuck at the `sorry` despite both sides looking identical:
```
∑ ij ∈ Finset.filter (fun ij ↦ ¬ij.IsDiag) f.support.sym2,
    Sym2.lift ⟨fun i j ↦ polar (⇑Q) (g i (f i)) (g j (f j)), ⋯⟩ ij =
  ∑ ij ∈ f.sym2OffDiag, Sym2.lift ⟨fun i j ↦ polar (⇑Q) (g i (f i)) (g j (f j)), ⋯⟩ ij
```

Reply:
These things both look the same to me... but not to mathlib:
```
⊢ ∑ ij ∈ Finset.filter (fun ij ↦ ¬ij.IsDiag) f.support.sym2,
    Sym2.lift ⟨fun i j ↦ polar (⇑Q) (g i (f i)) (g j (f j)), ⋯⟩ ij =
  ∑ ij ∈ Finset.filter (fun ij ↦ ¬ij.IsDiag) f.support.sym2,
    Sym2.lift ⟨fun i j ↦ polar (⇑Q) (g i (f i)) (g j (f j)), ⋯⟩ ij
```

Reply:
I strongly recommend not shooting yourself in the foot with `noncomputable instance : DecidableEq ι := by exact Classical.typeDecidableEq ι`

Reply:
> Would that be something of the form `def Finsupp.sym2OffDiag (f : ι →₀ R) : Finset (ι × ι) := sorry?`

No, I think I'm actually suggesting something more like
```
noncomputable def Finsupp.sym2 [CommMonoidWithZero R] (f : ι →₀ R) : Sym2 ι →₀ R :=
  .onFinset
    f.support.sym2
    (Sym2.lift ⟨fun i j => f i * f j, fun _ _ => mul_comm _ _⟩)
    (Sym2.ind <| by aesop)
```

Reply:
@eric-wieser presumably I want to sum over something like:
```
f.sym2Mul.support.filter (fun t => ¬ t.IsDiag)
```
But this gives me:
```
failed to synthesize
  DecidablePred fun t ↦ ¬t.IsDiag
```
Presumably
```
noncomputable instance : DecidablePred fun (t : Sym2 ι) ↦ ¬t.IsDiag := Classical.decPred _
```
isn't the correct solution?

Reply:
I guess I need `[DecidableEq ι]`?

Reply:
Okay, how about:
```
variable {ι R M N}

variable [CommRing R] [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N] [DecidableEq ι]

open Finsupp in
theorem map_finsupp_sum (Q : QuadraticMap R M N) (f : ι →₀ R) (g : ι → R → M) :
    Q (f.sum g) = (f.sum fun i r => Q (g i r)) +
    ∑ p ∈ f.support.sym2 with ¬ p.IsDiag,
      Sym2.lift
        ⟨fun i j => (polar Q) (g i (f i)) (g j (f j)), fun i j => by simp only [polar_comm]⟩ p := by
  rw [sum, QuadraticMap.map_sum]
  exact congrArg (HAdd.hAdd _) rfl
```